File: python/lesson04/solution/publisher.py
from flask import Flask, request
from opentelemetry import propagate, trace
from opentelemetry.sdk.trace import TracerProvider, Tracer
from lib.tracing import init_tracer_provider

app = Flask(__name__)

@app.route("/publish")
def publish():
    # obtain a tracer instance
    tracer: Tracer = get_tracer("publisher-tracer")
    
    # uncomment the following line to see the request headers
    # print(f"carrier: {request.headers} \nType: {type(request.headers)}")

    # extracting the span context from the request headers
    ctx = propagate.extract(carrier=request.headers)
    
    # print("context:", ctx)

    # starting a new span for the request
    with tracer.start_as_current_span('publish', context=ctx) as span:
        hello_str = request.args.get('helloStr')
        print(hello_str)
        # adding an event to the span indicating that request for publishing has been completed
        span.add_event('print-event', {'println': True})
        print(span.get_span_context())
        return "Published"

def get_tracer(tracer_name: str):
    # retrieve the global tracer provider
    tracer_provider: TracerProvider = trace.get_tracer_provider()
    # obtain a tracer instance from the tracer provider
    tracer: Tracer = tracer_provider.get_tracer(tracer_name)
    return tracer

def main():
    # setting and retrieving the global tracer provider
    tracer_provider: TracerProvider = init_tracer_provider('publisher')
    # starting the flask app on port 8082
    app.run(port=8082)
    # shutting down the tracer provider to ensure all the spans are exported
    tracer_provider.shutdown()

if __name__ == "__main__":
    main()

File: rust/lesson03/solution/src/bin/publisher.rs
use std::collections::HashMap;
use axum::{
    extract::Query,
    http::HeaderMap,
    routing::get,
    Router,
};
use tokio::net::TcpListener;

use opentelemetry::{
    global, trace::{Span, Tracer},
};
use opentelemetry_http::HeaderExtractor;

use exercise::init_tracer;


async fn publish_handler(Query(params): Query<HashMap<String, String>>,  headers: HeaderMap) {
    // creating a named instance of Tracer via the configured GlobalTracerProvider
    let tracer = global::tracer("publisher-tracer");
    
    // extracting the span context from the request headers
    let cx = global::get_text_map_propagator(|propagator| {
        propagator.extract(&HeaderExtractor(&headers))
    });
    
    // starting a new span named "publish" as a child of the extracted span context
    let mut span = tracer.start_with_context("publish", &cx);

    if let Some(hello_str) = params.get("hello_str") {
        println!("{}", hello_str);
    }
    span.end();
}



#[tokio::main]
async fn main() {

    // initializing the OpenTelemetry TracerProvider with the service name "publisher"
    let tp = init_tracer("publisher")
        .expect("Error initializing tracer");

    let app = Router::new().route("/publish", get(publish_handler));
    
    let listener = TcpListener::bind("0.0.0.0:8082").await.unwrap();
    
    let _ = axum::serve(listener, app).await;
    
    // shutting down the tracer provider to ensure all spans are flushed.
    tp.shutdown().expect("TracerProvider should shutdown successfully");

}